core/sup_tree_core/version_upgrade_task_queue.ex

```diff
@@ -14,17 +14,30 @@ defmodule AntikytheraCore.VersionUpgradeTaskQueue do
   use GenServer
   alias Antikythera.GearName
   alias AntikytheraCore.Version
+  require AntikytheraCore.Logger, as: L
 
   def start_link() do
     GenServer.start_link(__MODULE__, :ok, [name: __MODULE__])
   end
 
   @impl true
   def init(:ok) do
-    {:ok, %{queue: :queue.new(), task_pid: nil}}
+    {:ok, %{queue: :queue.new(), task_pid: nil, enable_upgrade?: true}}
   end
 
   @impl true
+  def handle_cast(message, state) when message in [:enable_version_upgrade, :disable_version_upgrade] do
+    new_state =
+      case message do
+        :enable_version_upgrade  ->
+          L.info("All upgrade processing of antikythera instance or gears is set again to enable.")
+          %{state | enable_upgrade?: true}
+        :disable_version_upgrade ->
+          L.info("All upgrade processing of antikythera instance or gears is set to disable.")
+          %{state | enable_upgrade?: false}
+      end
```

Comment:
Try something like the following:
```ex
def handle_cast(message, state) do
  new_state =
    case message do
      {:upgrade, instruction}  -> enqueue_message(state, instruction) |> run_task_if_possible()
      {:set_enabled, enabled?} -> set_enabled(state, enabled?)
    end
  {:noreply, new_state}
end

defp set_enabled(state, enabled?) do
  case {state.enabled?, enabled?} do
    {true , false} -> L.info("upgrade processing is turned off")
    {false, true } -> L.info("upgrade processing is turned on")
    _              -> :ok
  end
  %{state | enabled?: enabled?}
end
```

- Field name in process state should be `:enabled?` or `:upgrade_enabled?` rather than `:enable_upgrade?`.
- Now this `GenServer` receives 4 types of cast messages. Let's introduce a "structure" to classify them.

Reply:
It looks like a very smart code!

BTW, I came up with the implementation of "structure" below.
```
defmodule Message do
  use Croma.Struct, recursive_new?: true, fields: [
    upgrade:     Croma.TypeGen.nilable(Croma.Any),
    set_enabled: Croma.TypeGen.nilable(Croma.Boolean),
  ]
end
```
Is this data structure appropriate? Or is there a more appropriate data structure?

Reply:
Here what we want is a "tagged union" (or variant).
In Erlang/Elixir world the simplest way to implement "tagged union" is to use tuple.

I'm not a big fan of a struct with 2 nilable fields here, because e.g.
`%Message{upgrade: nil, set_enabled: nil}` is treated as a valid value for the type.
In order to circumvent the issue we could override `valid?/1` and `new/1`, but it's not worth the effort.

Note that the following things are basically orthogonal to the choice of data structure:
- whether to declare the type or not
- whether to validate arguments (using `defun` family) or not

So here (I think) simply using tuples is the way to go.

Reply:
Oh, I had a big misunderstanding.
I thought tuple and "structure" are different things. .
Thank you.

fixed in 32812e9